File: javascript/parse.js
const xml2js = require('xml2js')
const fs = require('fs');
const yargs = require('yargs');
const { DOMParser, XMLSerializer } = require('xmldom');
const xpath = require('xpath');


// it gives the result as an object
const readAndParseFile = (filePath) => {

    const data = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' });

    const xmlPromise = new Promise((res, rej) => {
        xml2js.parseString(data, (err, result) => {
            if (err) {
                rej(err)
            }
    
            res(result);        
        });    
    });

    return xmlPromise;
}

const writeFile = (json_object, file_path) => {

    const builder = new xml2js.Builder();

    const xmlString = builder.buildObject(json_object);

    fs.writeFile(file_path, xmlString, (err) => {
        if(err)
            console.log('unable to write the file in node js');
    })
}

const getParsingArgs = () => {

    const args = yargs
    .option('input', {
        description: 'Input XML file',
        type: 'string',
        demandOption: true
    })
    .option('output', {
        description: 'Output file for results',
        type: 'string',
        demandOption: true,
    })
    .option('query', {
        alias: 'q',
        description: 'XPath query',
        type: 'string',
        default: '',
    })
    .argv

    return [ args['input'], args['output'], args['q'] ];
}


const query_data = (xmlData, query) => {

    const xmlString = new xml2js.Builder().buildObject(xmlData);

    const doc = new DOMParser().parseFromString(xmlString);

    const serializer = new XMLSerializer();

    const nodes = xpath.select(query, doc);

    if(nodes.length == 0){
        console.log('EMPTY RESULT');
    }

    console.log(nodes.length)

    console.log("-----------------------Query Results---------------------------------");

    nodes.forEach(node => {
        const xmlOutput = serializer.serializeToString(node);
        console.log(xmlOutput);
    });

    console.log("-----------------------Query Results---------------------------------");
}

async function main() {
    const [ input_path, output_path, q ] = getParsingArgs();

    const result = await readAndParseFile(input_path);

    if(q)
        query_data(result, q);

    writeFile(result, output_path)
}


main();

File: python/parse.py

import xml.etree.ElementTree as ET
from lxml import etree
import sys
import argparse


def read_parse_file(file_path):
    with open(file_path, 'rb') as file:
            return etree.fromstring(file.read()) 

def write_file(path, xml_tree):
    with open(path, 'w', encoding='utf-8') as file:
        file.write(ET.tostring(xml_tree, encoding='unicode'))  

def get_parsing_args():
    args = sys.argv
    if len(args) < 3:
        raise Exception('Insufficient arguments provided')
    
    parser = argparse.ArgumentParser()
    parser.add_argument('--input', type=str, required=True, help='Input XML file')
    parser.add_argument('--output', type=str, required=True, help='Output file for results')
    parser.add_argument('-q', '--query', type=str, required=False, help='XPath query')

    args = parser.parse_args()

    args_obj = vars(args)

    input, output, query = args_obj['input'], args_obj['output'], args_obj['query']

    return input, output, query

def query_data(xml, query: str):
    query_result = xml.xpath(query)
    print(len(query_result))
    if(len(query_result) == 0):
        print()
        print("No Result")
        print()
        return

    print("-----------------------XPath Result------------------------")
    for index in range(len(query_result)):
        if isinstance(query_result[index], str):
            print( f"Element-{index}='{query_result[index]}'\n")
        else:
            print(
                f"Element-{index}='{etree.tostring(query_result[index], encoding='unicode')}'\n"
            )
    print("-----------------------------------------------------------")


def main():
    input, output, query = get_parsing_args()

    xml_tree = read_parse_file(input)

    if query is not None:
        query_data(xml_tree, query)

    write_file(output, xml_tree)

if __name__=="__main__":
    main()


        


    